src/server/pfs/server/driver.go

```diff
@@ -1407,64 +1407,199 @@ func (d *driver) deleteCommit(ctx context.Context, commit *pfs.Commit) error {
+		repos := d.repos.ReadWrite(stm)
 
-	// Delete the scratch space for this commit
-	prefix, err := d.scratchCommitPrefix(ctx, commit)
-	if err != nil {
-		return err
-	}
-	_, err = d.etcdClient.Delete(ctx, prefix, etcd.WithPrefix())
-	if err != nil {
-		return err
-	}
+		// re-read CommitInfo inside txn
+		commits := d.commits(commit.Repo.Name).ReadWrite(stm)
+		commitInfo = &pfs.CommitInfo{}
+		if err := commits.Get(commitID, commitInfo); err != nil {
+			return err
+		}
+		deleteScratch = commitInfo.Finished == nil
 
-	// If this commit is the head of a branch, make the commit's parent
-	// the head instead.
-	branchInfos, err := d.listBranch(ctx, commit.Repo)
-	if err != nil {
-		return err
-	}
+		// Helper for deleting commits
+		deleteCommit := func(repo col.ReadWriteCollection, commit *pfs.Commit) error {
+			// Store commitInfo in 'deleted' and remove commit from etcd
+			commitInfo := &pfs.CommitInfo{}
+			if err := repo.Get(commit.ID, commitInfo); err != nil {
+				return err
+			}
+			deleted[commit.ID] = commitInfo
+			if err := repo.Delete(commit.ID); err != nil {
+				return err
+			}
 
-	for _, branchInfo := range branchInfos {
-		if branchInfo.Head != nil && branchInfo.Head.ID == commitInfo.Commit.ID {
-			if err := d.createBranch(ctx, branchInfo.Branch, commitInfo.ParentCommit, nil); err != nil {
+			// Update repo size
+			// TODO this is basically wrong. Other commits could share data with the
+			// commit being removed, in which case we're subtracting too much. Doing
+			// the right thing is hard, though. Hopefully nobody ends up with a
+			// negative repo size and cares too much
+			repoInfo := &pfs.RepoInfo{}
+			if err := repos.Get(commit.Repo.Name, repoInfo); err != nil {
+				return err
+			}
+			repoInfo.SizeBytes -= commitInfo.SizeBytes
+			return repos.Put(commit.Repo.Name, repoInfo)
+		}
+
+		// Delete the commit itself and subtract the size of the commit from repo
+		// size.
+		deleteCommit(commits, commit)
+
+		// Delete all of the downstream commits of 'commit'
+		for _, subvCommit := range commitInfo.Subvenance {
+			downstreamRepo := d.commits(subvCommit.Repo.Name).ReadWrite(stm)
+			deleteCommit(downstreamRepo, subvCommit)
+		}
+
+		// Remove the commits in 'deleted' from all upstream commits' subvenance
+		for _, provCommit := range commitInfo.Provenance {
+			provCI := &pfs.CommitInfo{}
+			upstreamRepo := d.commits(provCommit.Repo.Name).ReadWrite(stm)
+			if err := upstreamRepo.Get(provCommit.ID, provCI); err != nil {
+				return err
+			}
+			// Remove deleted commits from provCI.Subvenance (copy 'from' -> 'to')
+			to := 0
+			for from, c := range provCI.Subvenance {
+				if _, ok := deleted[c.ID]; ok {
+					continue
+				}
+				provCI.Subvenance[to] = provCI.Subvenance[from]
+				to++
+			}
+			provCI.Subvenance = provCI.Subvenance[:to]
+			// Write updated commitInfo back to etcd
+			if err := upstreamRepo.Put(provCommit.ID, provCI); err != nil {
```

Comment:
How does this work with commit ranges? I think we might want to simplify things a little bit and say you can't delete commits that have provenance, you have to delete them through their provenance instead. I feel like there are some weird situations we can get into where you delete a commit and then pps just recreates it.

Reply:
I don't know if I totally understand this comment yet, but just as a quick note: PPS doesn't create commits anymore, so I actually think deleting a commit with provenance should be stable, even if it's also conceptually invalid. One problem you might have is a branch with a HEAD commit where the branch provenance and the HEAD commit provenance don't match (e.g. branch has 3 inputs, HEAD commit has 2). But this can happen even if you prevent people from deleting commits with provenance

Reply:
I understand the commit ranges comment now. I haven't merged that PR yet (so that will add some work to the end of this PR) but I agree that it'll make the code simpler and less risky if we add this constraint, even if it ends up being easy to remove, so I'll add it

Reply:
Ahh, good point about why it's stable. Sounds like we're in agreement about not being able to delete subvenant commits making things simpler. I also think this is better from a user experience perspective. Suppose I have a pipeline that outputs some garbage data which crashes a downstream pipeline, I might think that a good way to solve this is to delete that outputted garbage data, but the data that initial caused it to output garbage is still committed to the input repo so the next time I commit to it the garbage is going to come right back. Forcing people to only delete input commits forces them to solve the root problem.

Reply:
Makes sense. Just to update this thread, since it has to do with updating subvenance, I actually don't think there's a great way to avoid slightly gnarly subvenance code here. I was trying to figure out if two commits' subvenance always share a boundary (i.e. A.subvenance[i].upper == B.subvenance[j].lower), and one case I came up with that illustrates the need to rewrite subvenance (and which I'll add another test for) is this:
```
repo_1: repo_2: repo_3:
------  ------  ------
A       X       U
B       Y       V

output:
BVY
BVX
BUX
AUX
```

Note that if I delete `B`, then I have to delete `BUX` through `BVY`, which means updating the subvenance of commit `X`, even though both `X.subvenance[0].lower` and `X.subvenance[0].upper` don't equal `BUX` or `BVY`

Reply:
(One invariant I did come up with, which helps with drawing these diagrams: every output commit is the `lower` of a range in some upstream commit's subvenance.

This is because output commits are only created when new inputs commits are created, so the input commit that caused the new output commit will have the output commit as a `lower` somewhere)